File: sdk/python/zeroid/credential/verifier.py
"""Credential verification — verify VCs for status, expiry, and schema.

Provides the CredentialVerifier class which checks credential validity
including signature structure, expiration, status, and schema compliance.
"""

from __future__ import annotations

import hmac
from dataclasses import dataclass, field
from datetime import datetime, timezone

from zeroid.credential.proof import (
    build_credential_proof_payload,
    compute_credential_proof_value,
    compute_credential_subject_merkle_root,
)
from zeroid.credential.schema import SchemaRegistry
from zeroid.credential.types import CredentialStatus, VerifiableCredential


@dataclass(frozen=True)
class VerificationResult:
    """Result of a credential verification.

    Attributes:
        valid: Whether the credential passed all checks.
        errors: List of error messages for failed checks.
        warnings: List of warning messages.
    """

    valid: bool
    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)


class CredentialVerifier:
    """Verifies Verifiable Credentials.

    Checks credential structure, status, expiration, and optionally
    validates against a schema registry.
    """

    def __init__(
        self,
        schema_registry: SchemaRegistry | None = None,
        trusted_issuer_keys: dict[str, str] | None = None,
    ) -> None:
        """Initialize the verifier.

        Args:
            schema_registry: Optional schema registry for subject validation.
            trusted_issuer_keys: Mapping of issuer DID to verification key.
        """
        self.schema_registry = schema_registry
        self.trusted_issuer_keys = trusted_issuer_keys or {}

    def verify(self, credential: VerifiableCredential) -> VerificationResult:
        """Verify a Verifiable Credential.

        Checks:
        - Credential has required fields (id, issuer, issuance_date)
        - Credential status is ACTIVE
        - Credential has not expired
        - Proof is present and well-formed
        - Schema validation (if schema_registry and credential_schema set)

        Args:
            credential: The credential to verify.

        Returns:
            A VerificationResult with validity and error details.
        """
        errors: list[str] = []
        warnings: list[str] = []

        # Structure checks
        if not credential.id:
            errors.append("Missing credential ID")
        if not credential.issuer:
            errors.append("Missing issuer")
        if not credential.issuance_date:
            errors.append("Missing issuance date")

        # Status check
        if credential.credential_status == CredentialStatus.REVOKED:
            errors.append("Credential has been revoked")
        elif credential.credential_status == CredentialStatus.SUSPENDED:
            warnings.append("Credential is suspended")
        elif credential.credential_status == CredentialStatus.EXPIRED:
            errors.append("Credential is marked as expired")

        # Expiration check
        if credential.expiration_date:
            try:
                exp = datetime.fromisoformat(credential.expiration_date)
                if exp.tzinfo is None:
                    exp = exp.replace(tzinfo=timezone.utc)
                if exp < datetime.now(timezone.utc):
                    errors.append("Credential has expired")
            except ValueError:
                errors.append("Invalid expiration date format")

        # Proof checks
        if not credential.proof:
            errors.append("Missing proof")
        else:
            proof_value = credential.proof.get("proofValue")
            verification_method = credential.proof.get("verificationMethod")
            proof_type = credential.proof.get("type")
            proof_purpose = credential.proof.get("proofPurpose")
            merkle_root = credential.proof.get("merkleRoot")

            if not proof_value:
                errors.append("Missing proof value")
            if not verification_method:
                errors.append("Missing verification method in proof")
            elif not isinstance(verification_method, str):
                errors.append("Invalid verification method in proof")
            elif not verification_method.startswith(f"{credential.issuer}#"):
                errors.append("Verification method is not controlled by issuer")
            if proof_type != "ZeroIDCredentialProof2026":
                errors.append("Unsupported proof type")
            if proof_purpose != "assertionMethod":
                errors.append("Unsupported proof purpose")
            if not isinstance(merkle_root, str):
                errors.append("Missing subject Merkle root in proof")
            elif merkle_root != compute_credential_subject_merkle_root(
                credential.credential_subject
            ).hex():
                errors.append("Credential subject does not match proof root")

            issuer_key = self.trusted_issuer_keys.get(credential.issuer)
            if not issuer_key:
                errors.append("No trusted verification key configured for issuer")
            elif isinstance(proof_value, str) and isinstance(merkle_root, str):
                payload = build_credential_proof_payload(
                    credential_id=credential.id,
                    credential_types=credential.type,
                    issuer=credential.issuer,
                    issuance_date=credential.issuance_date,
                    expiration_date=credential.expiration_date,
                    credential_schema=credential.credential_schema,
                    subject_merkle_root=merkle_root,
                )
                expected = compute_credential_proof_value(issuer_key, payload)
                if not hmac.compare_digest(proof_value, expected):
                    errors.append("Credential proof signature is invalid")

        # Schema validation
        if credential.credential_schema and self.schema_registry:
            try:
                schema_errors = self.schema_registry.validate(
                    credential.credential_schema,
                    credential.credential_subject,
                )
                errors.extend(schema_errors)
            except KeyError:
                warnings.append(
                    f"Schema not found: {credential.credential_schema}"
                )

        return VerificationResult(
            valid=len(errors) == 0,
            errors=errors,
            warnings=warnings,
        )

File: sdk/go/credential/verifier.go
package credential

import (
	"crypto/subtle"
	"errors"
	"fmt"
	"strings"
	"time"
)

// Verification error sentinel values.
var (
	// ErrCredentialExpired is returned when a credential has passed its expiration date.
	ErrCredentialExpired = errors.New("credential: expired")
	// ErrCredentialRevoked is returned when a credential has been revoked.
	ErrCredentialRevoked = errors.New("credential: revoked")
	// ErrCredentialSuspended is returned when a credential has been suspended.
	ErrCredentialSuspended = errors.New("credential: suspended")
	// ErrCredentialInactive is returned when a credential has no active status.
	ErrCredentialInactive = errors.New("credential: inactive")
	// ErrIssuerNotApproved is returned when the issuer is not an approved issuer.
	ErrIssuerNotApproved = errors.New("credential: issuer not approved")
	// ErrNoProof is returned when a credential has no proof.
	ErrNoProof = errors.New("credential: no proof")
	// ErrInvalidProof is returned when a credential proof is malformed or invalid.
	ErrInvalidProof = errors.New("credential: invalid proof")
	// ErrIssuerKeyNotTrusted is returned when no trusted proof key is configured for the issuer.
	ErrIssuerKeyNotTrusted = errors.New("credential: issuer proof key not trusted")
	// ErrSchemaNotFound is returned when the credential schema is not found.
	ErrSchemaNotFound = errors.New("credential: schema not found")
)

// SchemaRegistry defines the interface for looking up credential schemas.
type SchemaRegistry interface {
	// GetSchema retrieves a credential schema by its ID.
	GetSchema(schemaID string) (*CredentialSchema, error)
}

// IssuerRegistry defines the interface for checking issuer authorization.
type IssuerRegistry interface {
	// IsApprovedIssuer checks whether a DID is an approved credential issuer.
	IsApprovedIssuer(did string) (bool, error)
}

// VerificationResult contains the result of credential verification.
type VerificationResult struct {
	// Valid indicates whether the credential passed all verification checks.
	Valid bool
	// Errors contains any verification errors encountered.
	Errors []string
	// Checks contains the individual check results.
	Checks map[string]bool
}

// Verifier verifies verifiable credentials against on-chain status,
// expiry, and issuer authorization.
type Verifier struct {
	schemas           SchemaRegistry
	issuers           IssuerRegistry
	trustedIssuerKeys map[string]string
	now               func() time.Time
}

// NewVerifier creates a new credential verifier with the given registries.
func NewVerifier(schemas SchemaRegistry, issuers IssuerRegistry) *Verifier {
	return &Verifier{
		schemas:           schemas,
		issuers:           issuers,
		trustedIssuerKeys: make(map[string]string),
		now:               time.Now,
	}
}

// NewVerifierWithTrustedIssuerKeys creates a verifier with issuer-scoped proof keys.
func NewVerifierWithTrustedIssuerKeys(schemas SchemaRegistry, issuers IssuerRegistry, keys map[string]string) *Verifier {
	v := NewVerifier(schemas, issuers)
	v.SetTrustedIssuerKeys(keys)
	return v
}

// SetTrustedIssuerKeys replaces the issuer DID to proof-key mapping used to
// authenticate credential proofs. Keys are copied so callers can safely mutate
// their input map after configuration.
func (v *Verifier) SetTrustedIssuerKeys(keys map[string]string) {
	v.trustedIssuerKeys = make(map[string]string, len(keys))
	for issuer, key := range keys {
		v.trustedIssuerKeys[issuer] = key
	}
}

// SetTimeFunc overrides the time function used for expiry checks. This is
// primarily useful for testing.
func (v *Verifier) SetTimeFunc(fn func() time.Time) {
	v.now = fn
}

// Verify performs a full verification of a verifiable credential, checking
// status, expiry, issuer authorization, and schema compliance.
func (v *Verifier) Verify(cred *VerifiableCredential) (*VerificationResult, error) {
	if cred == nil {
		return nil, errors.New("credential: nil credential")
	}

	result := &VerificationResult{
		Valid:  true,
		Checks: make(map[string]bool),
	}

	// Check status
	v.checkStatus(cred, result)

	// Check expiry
	v.checkExpiry(cred, result)

	// Check proof exists
	v.checkProof(cred, result)

	// Check issuer authorization
	if err := v.checkIssuer(cred, result); err != nil {
		return nil, fmt.Errorf("credential: issuer check failed: %w", err)
	}

	// Check schema if specified
	if err := v.checkSchema(cred, result); err != nil {
		return nil, fmt.Errorf("credential: schema check failed: %w", err)
	}

	return result, nil
}

func (v *Verifier) checkStatus(cred *VerifiableCredential, result *VerificationResult) {
	switch cred.Status {
	case StatusActive:
		result.Checks["status"] = true
	case StatusRevoked:
		result.Valid = false
		result.Checks["status"] = false
		result.Errors = append(result.Errors, ErrCredentialRevoked.Error())
	case StatusSuspended:
		result.Valid = false
		result.Checks["status"] = false
		result.Errors = append(result.Errors, ErrCredentialSuspended.Error())
	case StatusExpired:
		result.Valid = false
		result.Checks["status"] = false
		result.Errors = append(result.Errors, ErrCredentialExpired.Error())
	default:
		result.Valid = false
		result.Checks["status"] = false
		result.Errors = append(result.Errors, ErrCredentialInactive.Error())
	}
}

func (v *Verifier) checkExpiry(cred *VerifiableCredential, result *VerificationResult) {
	if cred.ExpirationDate.IsZero() {
		result.Checks["expiry"] = true
		return
	}
	if v.now().After(cred.ExpirationDate) {
		result.Valid = false
		result.Checks["expiry"] = false
		result.Errors = append(result.Errors, ErrCredentialExpired.Error())
	} else {
		result.Checks["expiry"] = true
	}
}

func (v *Verifier) checkProof(cred *VerifiableCredential, result *VerificationResult) {
	if cred.Proof == nil {
		result.Valid = false
		result.Checks["proof"] = false
		result.Errors = append(result.Errors, ErrNoProof.Error())
		return
	}

	proofErrors := make([]string, 0)
	proof := cred.Proof

	if proof.Type != "ZeroIDCredentialProof2026" {
		proofErrors = append(proofErrors, "credential: unsupported proof type")
	}
	if proof.ProofPurpose != "assertionMethod" {
		proofErrors = append(proofErrors, "credential: unsupported proof purpose")
	}
	if proof.Created.IsZero() {
		proofErrors = append(proofErrors, "credential: missing proof creation time")
	}
	if proof.VerificationMethod == "" {
		proofErrors = append(proofErrors, "credential: missing verification method")
	} else if !strings.HasPrefix(proof.VerificationMethod, cred.Issuer+"#") {
		proofErrors = append(proofErrors, "credential: verification method is not controlled by issuer")
	}
	if proof.ProofValue == "" {
		proofErrors = append(proofErrors, "credential: missing proof value")
	}
	if proof.MerkleRoot == "" {
		proofErrors = append(proofErrors, "credential: missing subject Merkle root")
	}

	subjectRoot, err := ComputeCredentialSubjectMerkleRoot(cred.CredentialSubject)
	if err != nil {
		proofErrors = append(proofErrors, err.Error())
	} else if proof.MerkleRoot != "" && proof.MerkleRoot != subjectRoot {
		proofErrors = append(proofErrors, "credential: subject does not match proof root")
	}

	issuerKey := v.trustedIssuerKeys[cred.Issuer]
	if issuerKey == "" {
		proofErrors = append(proofErrors, ErrIssuerKeyNotTrusted.Error())
	} else if proof.ProofValue != "" && proof.MerkleRoot != "" {
		payload, err := BuildCredentialProofPayload(cred, proof.MerkleRoot)
		if err != nil {
			proofErrors = append(proofErrors, err.Error())
		} else {
			expected := ComputeCredentialProofValue(issuerKey, payload)
			if subtle.ConstantTimeCompare([]byte(proof.ProofValue), []byte(expected)) != 1 {
				proofErrors = append(proofErrors, "credential: proof signature is invalid")
			}
		}
	}

	if len(proofErrors) > 0 {
		result.Valid = false
		result.Checks["proof"] = false
		result.Errors = append(result.Errors, proofErrors...)
		return
	}

	result.Checks["proof"] = true
}

func (v *Verifier) checkIssuer(cred *VerifiableCredential, result *VerificationResult) error {
	approved, err := v.issuers.IsApprovedIssuer(cred.Issuer)
	if err != nil {
		return err
	}
	if !approved {
		result.Valid = false
		result.Checks["issuer"] = false
		result.Errors = append(result.Errors, ErrIssuerNotApproved.Error())
	} else {
		result.Checks["issuer"] = true
	}
	return nil
}

func (v *Verifier) checkSchema(cred *VerifiableCredential, result *VerificationResult) error {
	if cred.SchemaID == "" {
		result.Checks["schema"] = true
		return nil
	}
	schema, err := v.schemas.GetSchema(cred.SchemaID)
	if err != nil {
		if errors.Is(err, ErrSchemaNotFound) {
			result.Valid = false
			result.Checks["schema"] = false
			result.Errors = append(result.Errors, ErrSchemaNotFound.Error())
			return nil
		}
		return err
	}
	validator := NewSchemaValidator()
	if err := validator.Validate(cred, schema); err != nil {
		result.Valid = false
		result.Checks["schema"] = false
		result.Errors = append(result.Errors, err.Error())
	} else {
		result.Checks["schema"] = true
	}
	return nil
}
